### DISCUSSION TITLE
Copilot Extension Awareness

### DISCUSSION THREAD

**Discussion by MSeal** [confluentinc/vscode#176] [Ideas]
### Goals

1. Enable contextual awareness about extension resources
2. Autocomplete for code that can use resources in API call patterns
3. Chat awareness for contextual clues on Q&A patterns
4. Action enablement for commands the chatbot could call for the user

### Non-Goals

1. Avoid sending potentially sensitive data (e.g. content of topics)
2. Should not be ubiquitously enabled (use or integration should be controlable)
3. Making any new interfaces for copilot or other AI integrations (use existing interfaces with context awareness)

### Background

This has been frequently asked about in offline conversations about the extension as a quality of life improvement. It especially speaks to enabling new users to get a moderate confidence pattern match to use of resources they may be seeing for the first time. The existing copilot integration into VS Code does not have specific context awareness for extension abstractions.

### Proposal

Integrate with latest vscode API changes to support context injection from an extension.

Majority of configurations required are in package.json, extension.ts, and (new) chat/participant.ts

- reference: https://code.visualstudio.com/api/extension-guides/chat#develop-a-chat-extension requires [VSCode Insiders](https://code.visualstudio.com/insiders/) to test
- proposed APIs from https://github.com/microsoft/vscode-extension-samples/tree/main/chat-sample once https://github.com/microsoft/vscode/issues/199908 is no longer open / proposed APIs are merged into the main VSCode release, we can update this
- related subsequent open changes to vscode for specific support:
    - https://github.com/microsoft/vscode/issues/212995
    - https://github.com/microsoft/vscode/issues/213274

**Comment by shouples** [comment]
I'd like to start outlining some more initial ideas where a Copilot chat participant can be helpful beyond a basic "AI chat bot that talks about Confluent/Kafka" via initial/system prompt:

1. **Schema definition creation/updates**
    - This could be done using some of the existing `#file`/`#selection` chat variables already leveraged by Copilot when a user opens the JSON editor view of a single message. (Requires prep work for general schema management functionality first.)
2. **Message Viewer filtering**
    - As part of https://github.com/confluentinc/vscode/issues/171, a chat participant could have a dedicated slash-command and button-output for opening Message Viewer to a given cluster+topic, timeframe, and search string (or more)
3. **Sample event/message generation**
    - After we are able to offer some producer actions, this seems like the next logical step: "generate some sample data" while pointing at an existing topic and/or schema.

Curious what others think about offerings here!

**Comment by MSeal** [comment]
Some related news: [extensions-public-beta](https://github.blog/news-insights/product-news/enhancing-the-github-copilot-ecosystem-with-copilot-extensions-now-in-public-beta) with some examples [here](https://github.com/copilot-extensions/?utm_source=cassidoo&utm_medium=email&utm_campaign=show-me-your-friends-and-ill-show-you-your-future). Courtesy of @Cerchie

**Comment by shouples** [comment]
May be handy: https://github.com/microsoft/vscode-chat-extension-utils
(from https://code.visualstudio.com/updates/v1_96#_vscodechatextensionutils)
